llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -167,29 +183,45 @@ class YkIRWriter {
     serialiseString(toString(V));
   }
 
-  void serialiseOperand(Instruction *Parent, Value *V) {
+  void serialiseOperand(Instruction *Parent, ValueLoweringMap &VLMap,
+                        Value *V) {
     if (llvm::Constant *C = dyn_cast<llvm::Constant>(V)) {
       serialiseConstantOperand(Parent, C);
     } else if (Instruction *I = dyn_cast<Instruction>(V)) {
       // If an instruction defines the operand, it's a local variable.
-      serialiseLocalVariableOperand(I);
+      serialiseLocalVariableOperand(I, VLMap);
     } else {
       serialiseUnimplementedOperand(V);
     }
   }
 
   /// Does a naiave serialisation of an LLVM instruction by iterating over its
   /// operands and serialising them in turn.
-  void serialiseInstGeneric(Instruction *I, OpCode Opc) {
+  void serialiseInstGeneric(Instruction *I, ValueLoweringMap &VLMap,
+                            unsigned BBIdx, unsigned &InstIdx, OpCode Opc) {
     OutStreamer.emitSizeT(typeIndex(I->getType()));
     serialiseOpcode(Opc);
     OutStreamer.emitInt32(I->getNumOperands());
     for (Value *O : I->operands()) {
-      serialiseOperand(I, O);
+      serialiseOperand(I, VLMap, O);
+    }
+    if (!I->getType()->isVoidTy()) {
+      VLMap[I] = {BBIdx, InstIdx};
     }
+    InstIdx++;
+  }
+
+  void serialiseInst(Instruction *I, ValueLoweringMap &VLMap, unsigned BBIdx,
+                     unsigned &InstIdx) {
+// Macro to help dispatch to generic lowering.
+//
+// Note that this is unhygenic so as to make the call-sites readable.
+#define GENERIC_INST_SERIALISE(LLVM_INST, LLVM_INST_TYPE, YKIR_OPCODE)         \
```

Comment:
I know you will comment on this, so I'll just preempt this now :)

I started with all the relevant scope being passed in to the macro, but this made the calls to the macro hard to read. So I've made the macro unhygenic, implicitely capturing scope from `serialiseInst()`, and moved the macro def inside of the function so as to imply this.